homeassistant/components/linknlink/coordinator.py

```diff
@@ -0,0 +1,192 @@
+"""LinknLink Coordinator."""
+from __future__ import annotations
+
+from contextlib import suppress
+from datetime import timedelta
+from functools import partial
+import logging
+from typing import Any
+
+import linknlink as llk
+from linknlink.exceptions import (
+    AuthenticationError,
+    AuthorizationError,
+    ConnectionClosedError,
+    LinknLinkException,
+    NetworkTimeoutError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_MAC, CONF_TIMEOUT, CONF_TYPE, Platform
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import DEFAULT_PORT, DOMAIN, DOMAINS_AND_TYPES
+
+_LOGGER = logging.getLogger(__name__)
+SCAN_INTERVAL = 10
+
+
+class Coordinator(DataUpdateCoordinator[dict[str, bytes]]):
+    """In charge of get the data for a site."""
+
+    api: llk.Device
+
+    def __init__(self, hass: HomeAssistant, config: ConfigEntry) -> None:
+        """Initialize the data service."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=config.title,
+            update_interval=timedelta(seconds=SCAN_INTERVAL),
+        )
+        self.name = config.title
+        self.hass = hass
+        self.config = config
+        self.fw_version: int | None = None
+        self.authorized: bool | None = None
+        self.reset_jobs: list[CALLBACK_TYPE] = []
+
+    @property
+    def available(self) -> bool | None:
+        """Return True if the device is available."""
+        return self.api.auth()
+
+    def unload(self) -> None:
+        """Cancel jobs."""
+        while self.reset_jobs:
+            self.reset_jobs.pop()()
```

Comment:
You can use `config_entry.async_on_unload(......)` to pass in functions it should execute when it unloads. I think you can use it at some places

Reply:
That's cool